# IoT Network Security Data Preprocessing

**Domain:** IoT  
**Dataset:** Secure IoT and WSN Sensor Data  
**Sumber:** Kaggle  
**Jumlah Data:** 1.000 baris  
**Format:** CSV  
**Link Dataset:** https://www.kaggle.com/datasets/zara2099/secure-iot-and-wsn-sensor-data

## Anggota Kelompok
1. Marthin Jordan Patuha Rero Silalahi  - 23.K1.0021
2. Klaudius Mario Saputra               - 24.K1.0007
3. Juwanli Takalamingan                 - 24.K1.0020


In [35]:
import pandas as pd

from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler

In [36]:
from google.colab import files

uploaded = files.upload()

Saving IoT_Sensor_Security_Dataset.csv to IoT_Sensor_Security_Dataset (1).csv


In [37]:
df = pd.read_csv("IoT_Sensor_Security_Dataset.csv")

df.head()

,temperature,humidity,pressure,light_intensity,sound_level,packet_loss,signal_strength,latency,throughput,jitter,failed_logins,unauthorized_access_attempts,malicious_packet_count,anomaly_score,threat_level
0,24.363503,32.959305,976.170568,672.702994,81.479629,1.968178,-54.622013,8.721091,72.054766,45.678901,6,4,1,0.120400,0
1,38.767858,57.933066,974.697880,796.681397,102.488910,2.367178,-87.932955,38.167733,68.759572,26.268024,3,0,12,0.122790,2
2,33.299849,81.106209,1040.625458,250.467899,98.414484,4.272737,-38.932381,166.417915,9.665844,36.245514,1,4,3,0.565303,1
3,29.966462,71.255742,974.954620,624.874100,43.850991,1.700022,-57.081863,153.586903,92.264983,21.802415,4,4,8,0.270096,1
4,18.900466,76.459280,977.194973,571.745983,43.432452,4.348248,-88.995728,70.777896,56.890373,31.501729,2,2,4,0.392399,1


### Rasionalisasi Acquisition & Ingestion

Dataset diperoleh dari Kaggle dalam format CSV dan terdiri dari 1.000 baris,
sehingga memenuhi ketentuan jumlah data yaitu antara 500 hingga 10.000 baris.
File CSV dimuat menggunakan library Pandas karena format data bersifat tabular
dan dapat langsung direpresentasikan sebagai DataFrame untuk proses audit dan
preprocessing selanjutnya.

In [38]:
print("Jumlah baris:", df.shape[0])
print("Jumlah kolom:", df.shape[1])

Jumlah baris: 1000
Jumlah kolom: 15


In [39]:
df.shape

(1000, 15)

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   temperature                   1000 non-null   float64
 1   humidity                      1000 non-null   float64
 2   pressure                      1000 non-null   float64
 3   light_intensity               1000 non-null   float64
 4   sound_level                   1000 non-null   float64
 5   packet_loss                   1000 non-null   float64
 6   signal_strength               1000 non-null   float64
 7   latency                       1000 non-null   float64
 8   throughput                    1000 non-null   float64
 9   jitter                        1000 non-null   float64
 10  failed_logins                 1000 non-null   int64  
 11  unauthorized_access_attempts  1000 non-null   int64  
 12  malicious_packet_count        1000 non-null   int64  
 13  anom

In [41]:
df.dtypes

,0
temperature,float64
humidity,float64
pressure,float64
light_intensity,float64
sound_level,float64
packet_loss,float64
signal_strength,float64
latency,float64
throughput,float64
jitter,float64


### Hasil Audit Format dan Schema

Hasil pemeriksaan menunjukkan bahwa dataset terdiri dari 1.000 baris dan
15 kolom. Sebanyak 14 kolom digunakan sebagai fitur dan satu kolom,
yaitu `threat_level`, digunakan sebagai target.

Berdasarkan hasil `df.info()` dan `df.dtypes`, fitur dataset telah memiliki
tipe data yang sesuai, yaitu `float64` dan `int64`. Oleh karena itu, tidak
diperlukan konversi tipe data tambahan sebelum preprocessing.

In [42]:
df.describe(include="all")

,temperature,humidity,pressure,light_intensity,sound_level,packet_loss,signal_strength,latency,throughput,jitter,failed_logins,unauthorized_access_attempts,malicious_packet_count,anomaly_score,threat_level
count,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000
mean,27.256414,55.491211,1000.240573,490.375048,74.469479,2.492322,-65.468381,96.991753,48.701020,24.839749,4.569000,2.040000,9.259000,0.500500,0.707000
std,7.303434,20.453293,29.067420,286.489344,25.812895,1.445974,20.281355,56.697965,28.195229,14.107662,2.832536,1.395115,5.904302,0.279737,0.779589
min,15.115801,20.225278,950.001163,0.653391,30.002765,0.030919,-99.900016,1.047960,0.124269,0.002641,0.000000,0.000000,0.000000,0.000119,0.000000
25%,20.899332,36.875199,976.135098,241.900770,52.047661,1.246900,-83.184484,47.961869,25.013433,13.207206,2.000000,1.000000,4.000000,0.265193,0.000000
50%,27.420184,56.311373,1000.061392,484.279284,74.513829,2.452014,-66.074988,97.314149,47.922225,24.759340,5.000000,2.000000,9.000000,0.507302,1.000000
75%,33.607990,73.232554,1025.910353,737.540811,96.599411,3.699259,-47.047496,143.178784,71.618907,36.810862,7.000000,3.000000,15.000000,0.733591,1.000000
max,39.992942,89.958961,1049.782086,999.557703,119.797445,4.996752,-30.037752,199.782035,99.601677,49.975259,9.000000,4.000000,19.000000,0.999009,2.000000


In [43]:
df.isnull().sum()

,0
temperature,0
humidity,0
pressure,0
light_intensity,0
sound_level,0
packet_loss,0
signal_strength,0
latency,0
throughput,0
jitter,0


### Penanganan Missing Value

Hasil pemeriksaan menggunakan `df.isnull().sum()` menunjukkan bahwa tidak
terdapat missing value pada seluruh kolom. Oleh karena itu, proses imputasi
tidak diterapkan karena seluruh observasi sudah memiliki nilai yang lengkap.
Melakukan imputasi pada kondisi ini tidak diperlukan dan tidak memberikan
manfaat terhadap kualitas data.

In [44]:
print("Jumlah duplikat:", df.duplicated().sum())

Jumlah duplikat: 0


### Pemeriksaan Data Duplikat

Hasil pemeriksaan menunjukkan bahwa tidak ditemukan baris data yang
terduplikasi. Oleh karena itu, tidak dilakukan penghapusan data duplikat
agar seluruh 1.000 observasi asli tetap dipertahankan.

In [45]:
# Memisahkan fitur dan target
X = df.drop(columns=["threat_level"])
y = df["threat_level"]

print("Jumlah fitur:", X.shape[1])
print("Jumlah data:", X.shape[0])

Jumlah fitur: 14
Jumlah data: 1000


## Penentuan Fitur dan Target

Dataset terdiri dari 14 variabel prediktor dan satu variabel target.
Kolom `threat_level` dipilih sebagai target klasifikasi karena
merepresentasikan tingkat ancaman pada data keamanan IoT, sedangkan
14 kolom lainnya digunakan sebagai fitur prediktor.

In [46]:
print(y.value_counts())

threat_level
0    493
1    307
2    200
Name: count, dtype: int64


In [47]:
print(y.value_counts(normalize=True) * 100)

threat_level
0    49.3
1    30.7
2    20.0
Name: proportion, dtype: float64


## Transformasi Fitur

Berdasarkan hasil audit tipe data, seluruh 14 variabel prediktor pada dataset
memiliki tipe data numerik (`float64` dan `int64`). Tidak ditemukan fitur
kategorikal nominal maupun ordinal, sehingga One-Hot Encoding (OHE) dan
Ordinal Encoding tidak diterapkan. Penerapan encoding pada fitur numerik
tidak diperlukan dan dapat mengubah makna asli data.

Seluruh fitur numerik kemudian distandardisasi menggunakan `StandardScaler`
melalui `ColumnTransformer`. Standardisasi dilakukan karena setiap fitur
memiliki rentang nilai dan satuan yang berbeda, seperti temperature,
pressure, latency, throughput, dan anomaly_score. Standardisasi membuat
fitur berada pada skala yang sebanding sehingga lebih siap digunakan pada
tahap pemodelan machine learning berikutnya.

In [48]:
numeric_features = X.columns.tolist()

print("Jumlah fitur numerik:", len(numeric_features))
print(numeric_features)

Jumlah fitur numerik: 14
['temperature', 'humidity', 'pressure', 'light_intensity', 'sound_level', 'packet_loss', 'signal_strength', 'latency', 'throughput', 'jitter', 'failed_logins', 'unauthorized_access_attempts', 'malicious_packet_count', 'anomaly_score']


In [49]:
preprocessor = ColumnTransformer(
    transformers=[
        ("num", StandardScaler(), numeric_features)
    ]
)

In [50]:
X_scaled = preprocessor.fit_transform(X)

print("Ukuran data setelah preprocessing:", X_scaled.shape)

Ukuran data setelah preprocessing: (1000, 14)


In [51]:
X_scaled[:5]

array([[-0.39630103, -1.10217857, -0.82848938,  0.63673984,  0.2717114 ,
        -0.36266703,  0.53506264, -1.55763667,  0.82870161,  1.47789048,
         0.50545372,  1.40560491, -1.39951054, -1.3594572 ],
       [ 1.57695733,  0.11944663, -0.8791793 ,  1.06970686,  1.08602506,
        -0.08658991, -1.10820082, -1.03801704,  0.71177253,  0.10129174,
        -0.55419768, -1.46297654,  0.46447008, -1.35090941],
       [ 0.82789256,  1.25299211,  1.39004747, -0.83782241,  0.92810149,
         1.23190691,  1.30904858,  1.22510395, -1.38515304,  0.80888469,
        -1.26063195,  1.40560491, -1.06060497,  0.23177301],
       [ 0.37125061,  0.77114327, -0.87034231,  0.46970805, -1.18676377,
        -0.54820968,  0.41371567,  0.99868608,  1.54585593, -0.21540452,
        -0.20098055,  1.40560491, -0.21334105, -0.82405824],
       [-1.14468466,  1.02568136, -0.79322939,  0.28416991, -1.20298622,
         1.2841549 , -1.16062855, -0.46257346,  0.29059708,  0.47246052,
        -0.90741482, -0.02

In [52]:
X_scaled_df = pd.DataFrame(
    X_scaled,
    columns=numeric_features
)

X_scaled_df.head()

,temperature,humidity,pressure,light_intensity,sound_level,packet_loss,signal_strength,latency,throughput,jitter,failed_logins,unauthorized_access_attempts,malicious_packet_count,anomaly_score
0,-0.396301,-1.102179,-0.828489,0.636740,0.271711,-0.362667,0.535063,-1.557637,0.828702,1.477890,0.505454,1.405605,-1.399511,-1.359457
1,1.576957,0.119447,-0.879179,1.069707,1.086025,-0.086590,-1.108201,-1.038017,0.711773,0.101292,-0.554198,-1.462977,0.464470,-1.350909
2,0.827893,1.252992,1.390047,-0.837822,0.928101,1.231907,1.309049,1.225104,-1.385153,0.808885,-1.260632,1.405605,-1.060605,0.231773
3,0.371251,0.771143,-0.870342,0.469708,-1.186764,-0.548210,0.413716,0.998686,1.545856,-0.215405,-0.200981,1.405605,-0.213341,-0.824058
4,-1.144685,1.025681,-0.793229,0.284170,-1.202986,1.284155,-1.160629,-0.462573,0.290597,0.472461,-0.907415,-0.028686,-0.891152,-0.386631


In [53]:
dataset_clean = pd.concat(
    [X_scaled_df, y.reset_index(drop=True)],
    axis=1
)

dataset_clean.head()

,temperature,humidity,pressure,light_intensity,sound_level,packet_loss,signal_strength,latency,throughput,jitter,failed_logins,unauthorized_access_attempts,malicious_packet_count,anomaly_score,threat_level
0,-0.396301,-1.102179,-0.828489,0.636740,0.271711,-0.362667,0.535063,-1.557637,0.828702,1.477890,0.505454,1.405605,-1.399511,-1.359457,0
1,1.576957,0.119447,-0.879179,1.069707,1.086025,-0.086590,-1.108201,-1.038017,0.711773,0.101292,-0.554198,-1.462977,0.464470,-1.350909,2
2,0.827893,1.252992,1.390047,-0.837822,0.928101,1.231907,1.309049,1.225104,-1.385153,0.808885,-1.260632,1.405605,-1.060605,0.231773,1
3,0.371251,0.771143,-0.870342,0.469708,-1.186764,-0.548210,0.413716,0.998686,1.545856,-0.215405,-0.200981,1.405605,-0.213341,-0.824058,1
4,-1.144685,1.025681,-0.793229,0.284170,-1.202986,1.284155,-1.160629,-0.462573,0.290597,0.472461,-0.907415,-0.028686,-0.891152,-0.386631,1


In [54]:
print("Jumlah baris:", dataset_clean.shape[0])
print("Jumlah kolom:", dataset_clean.shape[1])

Jumlah baris: 1000
Jumlah kolom: 15


In [55]:
dataset_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   temperature                   1000 non-null   float64
 1   humidity                      1000 non-null   float64
 2   pressure                      1000 non-null   float64
 3   light_intensity               1000 non-null   float64
 4   sound_level                   1000 non-null   float64
 5   packet_loss                   1000 non-null   float64
 6   signal_strength               1000 non-null   float64
 7   latency                       1000 non-null   float64
 8   throughput                    1000 non-null   float64
 9   jitter                        1000 non-null   float64
 10  failed_logins                 1000 non-null   float64
 11  unauthorized_access_attempts  1000 non-null   float64
 12  malicious_packet_count        1000 non-null   float64
 13  anom

In [56]:
dataset_clean.isnull().sum()

,0
temperature,0
humidity,0
pressure,0
light_intensity,0
sound_level,0
packet_loss,0
signal_strength,0
latency,0
throughput,0
jitter,0


In [57]:
dataset_clean["threat_level"].value_counts()

,count
threat_level,
0,493
1,307
2,200


In [58]:
dataset_clean.to_csv("dataset_clean.csv", index=False)

print("=== VALIDASI DATASET AKHIR ===")
print("Jumlah baris :", dataset_clean.shape[0])
print("Jumlah kolom :", dataset_clean.shape[1])
print("Missing value:", dataset_clean.isnull().sum().sum())
print("Duplikat     :", dataset_clean.duplicated().sum())

print("\nDistribusi target:")
print(dataset_clean["threat_level"].value_counts())
print("Dataset berhasil disimpan sebagai dataset_clean.csv")

files.download("dataset_clean.csv")

=== VALIDASI DATASET AKHIR ===
Jumlah baris : 1000
Jumlah kolom : 15
Missing value: 0
Duplikat     : 0

Distribusi target:
threat_level
0    493
1    307
2    200
Name: count, dtype: int64
Dataset berhasil disimpan sebagai dataset_clean.csv


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

### Hasil Validasi Dataset Akhir

Hasil validasi menunjukkan bahwa dataset hasil preprocessing terdiri dari
1.000 baris dan 15 kolom, yaitu 14 fitur prediktor dan satu variabel target
`threat_level`. Tidak terdapat missing value maupun data duplikat setelah
proses preprocessing.

Variabel `threat_level` memiliki tiga kelas, yaitu kelas 0 sebanyak 493 data
(49,3%), kelas 1 sebanyak 307 data (30,7%), dan kelas 2 sebanyak 200 data
(20,0%). Distribusi kelas tidak sepenuhnya seimbang, namun tidak dilakukan
resampling pada tahap ini karena fokus Milestone 1 adalah kurasi dan
preprocessing dataset. Ketidakseimbangan kelas dapat dipertimbangkan pada
tahap pemodelan dan evaluasi.

## Ringkasan Preprocessing

Dataset **Secure IoT and WSN Sensor Data** yang digunakan berasal dari Kaggle
dan terdiri dari 1.000 baris serta 15 kolom, sehingga memenuhi ketentuan
jumlah dataset yaitu 500–10.000 baris. Berdasarkan hasil audit schema,
terdapat 14 fitur prediktor dan satu variabel target yaitu `threat_level`.
Seluruh fitur telah memiliki tipe data numerik (`float64` dan `int64`)
sehingga tidak diperlukan konversi tipe data maupun encoding menggunakan
One-Hot Encoding atau Ordinal Encoding.

Hasil pemeriksaan menunjukkan bahwa dataset tidak memiliki missing value,
sehingga proses imputasi tidak diperlukan. Pemeriksaan data duplikat juga
dilakukan untuk memastikan tidak terdapat observasi yang berulang.
Seluruh fitur numerik kemudian distandardisasi menggunakan `StandardScaler`
melalui `ColumnTransformer` agar fitur yang memiliki satuan dan rentang nilai
berbeda berada pada skala yang sebanding. Variabel `threat_level` tidak
distandardisasi karena merupakan label target klasifikasi. Setelah seluruh
tahap preprocessing selesai, dataset akhir tetap terdiri dari 1.000 baris
dan 15 kolom dan disimpan dalam file `dataset_clean.csv`.